### PR TITLE
✨ Allow adding custom HA proxy config for CAPD load balancer

### PR DESCRIPTION
**Pull request by alexander-demicev**
**What this PR does / why we need it**:

Allow adding custom HA proxy config for CAPD load balancer. With this change, an additional load balancer configuration can be added using a config map. For more details see https://github.com/kubernetes-sigs/cluster-api/issues/7684

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes-sigs/cluster-api/issues/7684

### CODE REVIEW

File: test/infrastructure/docker/api/v1beta1/dockercluster_types.go [hunk 1]
```diff
@@ -53,6 +54,13 @@ type DockerClusterSpec struct {
 type DockerLoadBalancer struct {
 	// ImageMeta allows customizing the image used for the cluster load balancer.
 	ImageMeta `json:",inline"`
+
+	// AdditionalProxyConfigRef allows to add additional configuration to the HAProxy config file.
+	// This field is a reference to a config map that contains the extra configuration. The key of the config map
+	// should be equal to 'value'.
+	// The content of the config map will be appended to the HAProxy config file. Please use it with caution,
+	// as there are no checks to ensure the validity of the configuration.
+	AdditionalProxyConfigRef *corev1.LocalObjectReference `json:"additionalProxyConfigRef,omitempty"`
```
Comment (fabriziopandini) [inline test/infrastructure/docker/api/v1beta1/dockercluster_types.go:63]: What about
```suggestion
	UnsafeHAProxyConfigTemplateRef *corev1.LocalObjectReference `json:"additionalProxyConfigRef,omitempty"`
```
So we are explicity about "use at your own risk"; also, dropped additional according to latest discussion in the issue (it should be replace)

File: test/infrastructure/docker/internal/docker/loadbalancer.go [hunk 2]
```diff
@@ -161,16 +162,22 @@ func (s *LoadBalancer) UpdateConfiguration(ctx context.Context) error {
 			return errors.Wrapf(err, "failed to get IP for container %s", n.String())
 		}
 		if s.ipFamily == clusterv1.IPv6IPFamily {
-			backendServers[n.String()] = net.JoinHostPort(controlPlaneIPv6, "6443")
+			backendServers[n.String()] = controlPlaneIPv6
```
Comment (fabriziopandini) [inline test/infrastructure/docker/internal/docker/loadbalancer.go:165]: This change seems unrelated, could you provide some context?

File: test/infrastructure/docker/internal/docker/loadbalancer.go [hunk 3]
```diff
@@ -137,7 +138,7 @@ func (s *LoadBalancer) Create(ctx context.Context) error {
 }
 
 // UpdateConfiguration updates the external load balancer configuration with new control plane nodes.
-func (s *LoadBalancer) UpdateConfiguration(ctx context.Context) error {
+func (s *LoadBalancer) UpdateConfiguration(ctx context.Context, additionalLoadBalancerConfigCM *corev1.ConfigMap) error {
```
Comment (fabriziopandini) [inline test/infrastructure/docker/internal/docker/loadbalancer.go:141]: This should be updated according to latest discussions on the issue

File: test/infrastructure/docker/internal/docker/loadbalancer.go (hunk 2 above)
Comment (alexander-demicev) [reply test/infrastructure/docker/internal/docker/loadbalancer.go:165]: I might be wrong but it seemed like a bug to me, we are passing `controlPlanePort` as part of `LoadBalancer` object, and on these lines, it's hardcoded to "6443". I removed it and the port is later processed in the template.

Comment (alexander-demicev) [reply test/infrastructure/docker/internal/docker/loadbalancer.go:166]: I split the backend and frontend control plane port in the `Loadbalancer` object, with a custom template we might want the server address to contain only IP address without a port.
With this change, the port will be appended to the IP address later when processing the template.

File: test/infrastructure/docker/internal/loadbalancer/config.go [hunk 4]
```diff
@@ -18,16 +18,19 @@ package loadbalancer
...
 // ConfigData is supplied to the loadbalancer config template.
 type ConfigData struct {
-	ControlPlanePort int
-	BackendServers   map[string]string
-	IPv6             bool
+	FrontendControlPlanePort string
+	BackendControlPlanePort  string
+	BackendServers           map[string]string
+	IPv6                     bool
+	UnsafeLoadBalancerConfig string
```
Comment (richardcase) [inline test/infrastructure/docker/internal/loadbalancer/config.go:33]: This feels like it shouldn't live on this struct, as this struct is use when templating and contains data to be used within a template.

File: test/infrastructure/docker/internal/controllers/dockermachine_controller.go [hunk 5]
```diff
@@ -267,7 +268,16 @@ func (r *DockerMachineReconciler) reconcileNormal(ctx context.Context, cluster *
 	// we should only do this once, as reconfiguration more or less ensures
 	// node ref setting fails
 	if util.IsControlPlaneMachine(machine) && !dockerMachine.Status.LoadBalancerConfigured {
-		if err := externalLoadBalancer.UpdateConfiguration(ctx); err != nil {
+		additionalConfigCM := &corev1.ConfigMap{}
```
Comment (fabriziopandini) [inline test/infrastructure/docker/internal/controllers/dockermachine_controller.go:271]:
```suggestion
		var additionalConfigCM *corev1.ConfigMap{}
```
nit: So it will be nil in case it does not exists (vs being an empty config map); same for reconcileDelete

File: test/infrastructure/docker/internal/loadbalancer/config.go [hunk 6]
```diff
@@ -55,23 +57,30 @@ defaults
...
-  server {{ $server }} {{ $address }} check check-ssl verify none resolvers docker resolve-prefer {{ if $.IPv6 -}} ipv6 {{- else -}} ipv4 {{- end }}
+  server {{ $server }} {{ JoinHostPort $address $.BackendControlPlanePort }} check check-ssl verify none resolvers docker resolve-prefer {{ if $.IPv6 -}} ipv6 {{- else -}} ipv4 {{- end }}
   {{- end}}
 `
 
 // Config generates the loadbalancer config from the ConfigTemplate and ConfigData.
-func Config(data *ConfigData) (config string, err error) {
-	t, err := template.New("loadbalancer-config").Parse(ConfigTemplate)
+func Config(data *ConfigData, customTemplate string) (config string, err error) {
+	configTemplate := ConfigTemplate
```
Comment (fabriziopandini) [inline test/infrastructure/docker/internal/loadbalancer/config.go:76]: what about
```suggestion
	configTemplate := DefaultTemplate
```
(not sure it this is used somewhere else...)

**Review by fabriziopandini — COMMENTED** [review]
Just two nits from my side

File: test/infrastructure/docker/internal/loadbalancer/config.go (hunk 6 above)
Comment (alexander-demicev) [reply test/infrastructure/docker/internal/loadbalancer/config.go:76]: since it's in internal package, it should be safe to change the name

**Comment by fabriziopandini** [comment]
/lgtm
@chrischdi @elmiko PTAL

File: test/infrastructure/docker/internal/loadbalancer/config.go [hunk 7]
```diff
@@ -55,23 +57,30 @@ defaults
...
-	t, err := template.New("loadbalancer-config").Parse(ConfigTemplate)
+func Config(data *ConfigData, customTemplate string) (config string, err error) {
+	configTemplate := DefaultTemplate
+	if customTemplate != "" {
+		configTemplate = customTemplate
+	}
+
+	t, err := template.New("loadbalancer-config").Funcs(template.FuncMap{
+		"JoinHostPort": net.JoinHostPort,
+	}).Parse(configTemplate)
```
Comment (chrischdi) [inline test/infrastructure/docker/internal/loadbalancer/config.go:83]: Wondering if we should keep it simple in here and passing the `DefaultTemplate` already from the above func.

```suggestion
func Config(data *ConfigData, configTemplate string) (config string, err error) {
	t, err := template.New("loadbalancer-config").Funcs(template.FuncMap{
		"JoinHostPort": net.JoinHostPort,
	}).Parse(configTemplate)
```

File: test/infrastructure/docker/internal/docker/loadbalancer.go [hunk 8]
```diff
@@ -161,17 +164,25 @@ func (s *LoadBalancer) UpdateConfiguration(ctx context.Context) error {
...
 		} else {
-			backendServers[n.String()] = net.JoinHostPort(controlPlaneIPv4, "6443")
+			backendServers[n.String()] = controlPlaneIPv4
 		}
 	}
 
+	unsafeLoadBalancerConfig := ""
+	if unsafeLoadBalancerConfigCM != nil {
+		unsafeLoadBalancerConfig = unsafeLoadBalancerConfigCM.Data["value"]
+	}
```
Comment (chrischdi) [inline test/infrastructure/docker/internal/docker/loadbalancer.go:176]:
```suggestion
	loadBalancerConfigTemplate := loadbalancer.DefaultTemplate
	if unsafeLoadBalancerConfigCM != nil {
		loadBalancerConfigTemplate = unsafeLoadBalancerConfigCM.Data["value"]
	}
```

File: test/infrastructure/docker/internal/docker/loadbalancer.go [hunk 9]
```diff
@@ -137,7 +140,7 @@ func (s *LoadBalancer) Create(ctx context.Context) error {
 }
 
 // UpdateConfiguration updates the external load balancer configuration with new control plane nodes.
-func (s *LoadBalancer) UpdateConfiguration(ctx context.Context) error {
+func (s *LoadBalancer) UpdateConfiguration(ctx context.Context, unsafeLoadBalancerConfigCM *corev1.ConfigMap) error {
```
Comment (chrischdi) [inline test/infrastructure/docker/internal/docker/loadbalancer.go:143]: We could keep the whole configmap logic at the calling function and only pass a string as `unsafeLoadBalancerConfigTemplate` here. And use the template if it is `!= ""`

```suggestion
func (s *LoadBalancer) UpdateConfiguration(ctx context.Context, unsafeLoadBalancerConfigCM string) error {
```

This way all the logic is at the above function (getting configmap, configmap key).

File: test/infrastructure/docker/internal/controllers/dockermachine_controller.go [hunk 10]
```diff
@@ -267,7 +268,17 @@ func (r *DockerMachineReconciler) reconcileNormal(ctx context.Context, cluster *
...
+		if dockerCluster.Spec.LoadBalancer.UnsafeHAProxyConfigTemplateRef != nil {
+			unsafeLoadBalancerConfigCM = &corev1.ConfigMap{}
+			if err := r.Get(ctx, types.NamespacedName{
+				Name:      dockerCluster.Spec.LoadBalancer.UnsafeHAProxyConfigTemplateRef.Name,
+				Namespace: dockerMachine.Namespace,
+			}, unsafeLoadBalancerConfigCM); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, "failed to retrieve custom HAProxy configuration ConfigMap")
+			}
+		}
+		if err := externalLoadBalancer.UpdateConfiguration(ctx, unsafeLoadBalancerConfigCM); err != nil {
```
Comment (chrischdi) [inline test/infrastructure/docker/internal/controllers/dockermachine_controller.go:281]: Because this code is duplicated below: what about creating a helper function and using that one instead? (note, I paired the suggestion together with passing just a string to the `UpdateConfiguration` function)

```suggestion
		unsafeLoadBalancerConfigTemplate, err := getUnsafeLoadBalancerConfigTemplate(ctx)
		if err != nil {
				return ctrl.Result{}, errors.Wrap(err, "failed to retrieve HAProxy configuration from UnsafeHAProxyConfigTemplateRef")
		}
		if err := externalLoadBalancer.UpdateConfiguration(ctx, unsafeLoadBalancerConfigTemplate); err != nil {
```

```go
func (r *DockerMachineReconciler) getUnsafeLoadBalancerConfigTemplate (ctx context.Context, dockerCluster *infrav1.DockerCluster) (string, error) {
	if dockerCluster.Spec.LoadBalancer.UnsafeHAProxyConfigTemplateRef == nil {
		return "", nil
	}
	var cm *corev1.ConfigMap
	key := types.NamespacedName{
		Name:      dockerCluster.Spec.LoadBalancer.UnsafeHAProxyConfigTemplateRef.Name,
		Namespace: dockerMachine.Namespace,
	}
	if err := r.Get(ctx, key, cm ); err != nil {
		return "", errors.Wrapf(err, "failed to retrieve custom HAProxy configuration ConfigMap %s", key)
	}
	template, ok := unsafeLoadBalancerConfigCM.Data["value"]
	if !ok {
		return fmt.Errorf("Expected key \"value\" to exist in ConfigMap %s", key)
	}
	return template, nil
}
```

This also allows to return an error if the key does not exist.

**Review by chrischdi — COMMENTED** [review]
Some smaller nits :-)

**Comment by alexander-demicev** [comment]
@chrischdi Thanks! all fixed

File: test/infrastructure/docker/internal/controllers/dockermachine_controller.go [hunk 11]
```diff
@@ -510,6 +519,25 @@ func (r *DockerMachineReconciler) getBootstrapData(ctx context.Context, machine
 	return base64.StdEncoding.EncodeToString(value), bootstrapv1.Format(format), nil
 }
 
+func (r *DockerMachineReconciler) getUnsafeLoadBalancerConfigTemplate(ctx context.Context, dockerCluster *infrav1.DockerCluster, dockerMachine *infrav1.DockerMachine) (string, error) {
```
Comment (chrischdi) [inline test/infrastructure/docker/internal/controllers/dockermachine_controller.go:522]:
```suggestion
func (r *DockerMachineReconciler) getUnsafeLoadBalancerConfigTemplate(ctx context.Context, dockerCluster *infrav1.DockerCluster) (string, error) {
```

File: test/infrastructure/docker/internal/controllers/dockermachine_controller.go [hunk 12]
```diff
@@ -510,6 +519,25 @@ func (r *DockerMachineReconciler) getBootstrapData(ctx context.Context, machine
...
+	if dockerCluster.Spec.LoadBalancer.UnsafeHAProxyConfigTemplateRef == nil {
+		return "", nil
+	}
+	var cm *corev1.ConfigMap
+	key := types.NamespacedName{
+		Name:      dockerCluster.Spec.LoadBalancer.UnsafeHAProxyConfigTemplateRef.Name,
+		Namespace: dockerMachine.Namespace,
```
Comment (chrischdi) [inline test/infrastructure/docker/internal/controllers/dockermachine_controller.go:529]:
```suggestion
		Namespace: dockerCluster.Namespace,
```

File: test/infrastructure/docker/internal/controllers/dockermachine_controller.go [hunk 13]
```diff
@@ -411,7 +416,11 @@ func (r *DockerMachineReconciler) reconcileDelete(ctx context.Context, machine *
 
 	// if the deleted machine is a control-plane node, remove it from the load balancer configuration;
 	if util.IsControlPlaneMachine(machine) {
-		if err := externalLoadBalancer.UpdateConfiguration(ctx); err != nil {
+		unsafeLoadBalancerConfigTemplate, err := r.getUnsafeLoadBalancerConfigTemplate(ctx, dockerCluster, dockerMachine)
```
Comment (chrischdi) [inline test/infrastructure/docker/internal/controllers/dockermachine_controller.go:419]:
```suggestion
		unsafeLoadBalancerConfigTemplate, err := r.getUnsafeLoadBalancerConfigTemplate(ctx, dockerCluster)
```

File: test/infrastructure/docker/internal/controllers/dockermachine_controller.go [hunk 14]
```diff
@@ -271,7 +272,11 @@ func (r *DockerMachineReconciler) reconcileNormal(ctx context.Context, cluster *
 	// we should only do this once, as reconfiguration more or less ensures
 	// node ref setting fails
 	if util.IsControlPlaneMachine(machine) && !dockerMachine.Status.LoadBalancerConfigured {
-		if err := externalLoadBalancer.UpdateConfiguration(ctx); err != nil {
+		unsafeLoadBalancerConfigTemplate, err := r.getUnsafeLoadBalancerConfigTemplate(ctx, dockerCluster, dockerMachine)
```
Comment (chrischdi) [inline test/infrastructure/docker/internal/controllers/dockermachine_controller.go:275]:
```suggestion
		unsafeLoadBalancerConfigTemplate, err := r.getUnsafeLoadBalancerConfigTemplate(ctx, dockerMachine)
```

**Review by chrischdi — COMMENTED** [review]
One last nit but also ok to do it in a follow up :-) 

As of that lgtm 👍

File: test/infrastructure/docker/api/v1beta1/dockercluster_types.go [hunk 15]
```diff
@@ -53,6 +54,17 @@ type DockerClusterSpec struct {
...
+	// This field is a reference to a config map that contains the configuration template. The key of the config map should be equal to 'value'.
+	// The content of the config map will be processed and will replace the default HAProxy config file. Please use it with caution, as there are
+	// no checks to ensure the validity of the configuration. This template will support the following variables that will be passed by the controller:
+	// $IPv6 (bool) indicates if the cluster is IPv6, $FrontendControlPlanePort (string) indicates the frontend control plane port,
+	// $BackendControlPlanePort (string) indicates the backend control plane port, $BackendServers (map[string]string) indicates the backend server
+	// where the key is the server name and the value is the address. This map is dynamic and is updated every time a new control plane
+	// node is added or removed. The template will also support the JoinHostPort function to join the host and port of the backend server.
+	// +optional
+	UnsafeHAProxyConfigTemplateRef *corev1.LocalObjectReference `json:"unsafeHAProxyConfigTemplateRef,omitempty"`
```
Comment (richardcase) [inline test/infrastructure/docker/api/v1beta1/dockercluster_types.go:67]: Re-raising this as i think it got lost.

Not sure i like the use of `Unsafe` as that sounds scary :skull:  Could we use something like `CustomHAProxyConfigTemplateRef` or `CustomLoadBalancerConfigRef` ?

Comment (chrischdi) [reply test/infrastructure/docker/api/v1beta1/dockercluster_types.go:67]: No big opinions on my side either :-) I think the unsafe comes from:

> Let's [...] make sure to clearly document the contract about it: e.g. no validation, use at your own risk, etc.

[[0]](https://github.com/kubernetes-sigs/cluster-api/issues/7684#issuecomment-1565358197)

**Review by chrischdi — COMMENTED** [review]
/lgtm

pending the naming discussion :-)

File: test/infrastructure/docker/api/v1beta1/dockercluster_types.go (hunk 15 above)
Comment (richardcase) [reply test/infrastructure/docker/api/v1beta1/dockercluster_types.go:67]: There are a number of areas where there is no validation, and you can do lots of damage. For example  `PreKubeadmCommands`.  

If we have a warning in the documentation of the field that states "you better know what you are doing" then that's probably enough.

Comment (sbueringer) [reply test/infrastructure/docker/api/v1beta1/dockercluster_types.go:67]: I agree with Richard. I think we don't need Unsafe in the field name

There are so many ways to misconfigure Cluster API easily and it's totally impossible for us to validate. We can't prefix everything with Unsafe

**Comment by sbueringer** [comment]
/lgtm

pending getting consensus on the field name (no strong opinion from my side)

**Comment by richardcase** [comment]
Thanks @alexander-demicev for the name change :) For me:

/lgtm

**Comment by sbueringer** [comment]
Thank you very much!

/lgtm
/approve
